src/consumer_queue.h

```diff
@@ -2,60 +2,167 @@
-
+	consumer_queue& operator= (consumer_queue&&) = delete;
 private:
-	send_buffer_p registry_; // optional consumer registry
-	buffer_type buffer_;	 // the sample buffer
-	std::mutex mut_;			 // mutex for cond var (also to protect queue at buffer overflow)
-	std::condition_variable cv_; // to allow for blocking wait by consumer
+	// an item stored in the queue
+	struct item_t {
+		std::atomic<std::size_t> seq_state;
+		sample_p value;
+	};
+
+	// Push a new element to the queue.
+	// Returns true if successful or false if queue full.
+	template <class T>
+	bool try_push(T&& sample) {
+		std::size_t write_index = write_idx_.load(std::memory_order_acquire);
+		std::size_t next_idx = add_wrap(write_index, 1);
+		item_t &item = buffer_[write_index % size_];
+		if (UNLIKELY(write_index != item.seq_state.load(std::memory_order_acquire)))
+			return false;  // item currently occupied, queue full
+		write_idx_.store(next_idx, std::memory_order_release);
+		copy_or_move(item.value, std::forward<T>(sample));
+		item.seq_state.store(next_idx, std::memory_order_release);
+		return true;
+	}
+
+	// Pop an element from the queue (can be called with zero or one argument). Returns true if
+	// successful or false if queue is empty. Uses the same method as Vyukov's bounded MPMC queue.
+	template <class ... T>
+	bool try_pop(T&... result) {
+		item_t* item;
+		std::size_t read_index = read_idx_.load(std::memory_order_relaxed);
+		for (;;) {
+			item = &buffer_[read_index % size_];
+			const std::size_t seq_state = item->seq_state.load(std::memory_order_acquire);
+			const std::size_t next_idx = add_wrap(read_index, 1);
+			// check if the item is ok to pop
+			if (LIKELY(seq_state == next_idx)) {
+				// yes, try to claim slot using CAS
+				if (LIKELY(read_idx_.compare_exchange_weak(read_index, next_idx, std::memory_order_relaxed)))
+					break;
+			} else if (LIKELY(seq_state == read_index))
+				return false;  // queue empty
+			else
+				// we're behind or ahead of another pop, try again
+				read_index = read_idx_.load(std::memory_order_relaxed);
+		}
+		move_or_drop(item->value, result...);
+		// mark item as free for next pass
+		item->seq_state.store(add_wrap(read_index, size_), std::memory_order_release);
+		return true;
+	}
+
+	// helper to either copy or move a value, depending on whether it's an rvalue ref
+	inline static void copy_or_move(sample_p& dst, const sample_p& src) { dst = src; }
+	inline static void copy_or_move(sample_p& dst, sample_p&& src) { dst = std::move(src); }
+	// helper to either move or drop a value, depending on whether a dst argument is given
+	inline static void move_or_drop(sample_p &src) { }
+	inline static void move_or_drop(sample_p &src, sample_p &dst) { dst = std::move(src); }
+	// helper to add a delta to the given index and wrap correctly
+	FORCEINLINE std::size_t add_wrap(std::size_t x, std::size_t delta) const {
+		const std::size_t xp = x + delta;
+		return xp >= wrap_at_ ? xp - wrap_at_ : xp;
```

Comment:
I originally thought that an `idiv + subtract` (`(xp/wrap_at_)+xp%wrap_at`?) would be faster, but both x64 and ARM have conditional move / conditional subtract instructions.

Reply:
Yeah I had wondered the same thing and had tested a `return (x + delta) % wrap_at_`, and surprisingly it was quite a bit slower than the conditional move version. I guess the most important arch targets are those two, and archs that don't have `cmov` may also have an even slower modulo operation.